pyo3-macros-backend/src/introspection.rs

```diff
@@ -50,56 +53,137 @@ pub fn module_introspection_code<'a>(
         ]
         .into(),
     )
-    .emit(pyo3_crate_path);
-    let introspection_id = introspection_id_const();
-    quote! {
-        #stub
-        #introspection_id
-    }
+    .emit(pyo3_crate_path)
 }
 
 pub fn class_introspection_code(
     pyo3_crate_path: &PyO3CratePath,
     ident: &Ident,
     name: &str,
 ) -> TokenStream {
-    let stub = IntrospectionNode::Map(
+    IntrospectionNode::Map(
         [
-            ("type", IntrospectionNode::String("class")),
+            ("type", IntrospectionNode::String("class".into())),
             ("id", IntrospectionNode::IntrospectionId(Some(ident))),
-            ("name", IntrospectionNode::String(name)),
+            ("name", IntrospectionNode::String(name.into())),
         ]
         .into(),
     )
-    .emit(pyo3_crate_path);
-    let introspection_id = introspection_id_const();
-    quote! {
-        #stub
-        impl #ident {
-            #introspection_id
-        }
-    }
+    .emit(pyo3_crate_path)
 }
 
-pub fn function_introspection_code(pyo3_crate_path: &PyO3CratePath, name: &str) -> TokenStream {
-    let stub = IntrospectionNode::Map(
+pub fn function_introspection_code(
+    pyo3_crate_path: &PyO3CratePath,
+    ident: &Ident,
+    name: &str,
+    signature: &FunctionSignature<'_>,
+) -> TokenStream {
+    IntrospectionNode::Map(
         [
-            ("type", IntrospectionNode::String("function")),
-            ("id", IntrospectionNode::IntrospectionId(None)),
-            ("name", IntrospectionNode::String(name)),
+            ("type", IntrospectionNode::String("function".into())),
+            ("id", IntrospectionNode::IntrospectionId(Some(ident))),
+            ("name", IntrospectionNode::String(name.into())),
+            ("arguments", arguments_introspection_data(signature)),
         ]
         .into(),
     )
-    .emit(pyo3_crate_path);
-    let introspection_id = introspection_id_const();
-    quote! {
-        #stub
-        #introspection_id
+    .emit(pyo3_crate_path)
+}
+
+fn arguments_introspection_data<'a>(signature: &'a FunctionSignature<'a>) -> IntrospectionNode<'a> {
```

Comment:
I feel like the logic here might be somewhat duplicated with other code in the macros which sets up the argument parsing. Do you think there's opportunity to refactor and reuse?

Reply:
Let's keep this as-is for now, maybe we refactor later if there seems to be opportunity.